quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/health/ReactiveHealthHandler.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.quarkus.runtime.integration.health;
+
+import io.quarkus.smallrye.health.runtime.SmallRyeLivenessHandler;
+import io.smallrye.health.SmallRyeHealth;
+import io.smallrye.health.SmallRyeHealthReporter;
+import io.smallrye.mutiny.Uni;
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * This adds the possibility to have a non-blocking health handler in Quarkus.
+ * <p>
+ * Without a non-blocking health check, all liveness and readiness probes will enqueue in the worker thread pool. Under high load
```

Comment:
I totally love your added explanations here! I am wondering if what you describe here isn't a problem in itself: if Infinispan rebalancing competes with threads from incoming requests, that means lots of requests can make the cluster break. Shouldn't Infinispan have its own thread pool?

Reply:
Hi @sschu - thanks for commenting here. Infinispan has its own thread pool, and it works like this AFAIK: 

Keycloak / RESTEasy request: 
1. Incoming request is dispatched by Quarkus in the Vert.x nonblocking event loop
2. Handover of the request to RESTEasy, which uses for all blocking requests the Quarkus executor pool
3. If the request processing needs data from a remote Infinispan, it will reach out to the remote Infinispan. While doing this, the Quarkus executor thread is still blocked
4. The Infinispan processing will use its own threads in local and remote Infinispan  

During rebalancing or node failures, the Infinispan handling will take longer due to timeouts and retries, and therefore the Quarkus executor threads will be blocked for a longer time. When the number of in-flight requests reaches the number of executor pool threads, the requests will be queued. 

As the default, Quarkus probes (as well as metrics) travel the same route: After being dispatched by the Vert.x event loop first, they are then handled by blocking handlers which run on the Quarkus executor pool. If they are queued, their results will exceed the Kubernetes probe timeout.

This PR updates the readiness and liveness probes to use nonblocking handlers. If then all checks of the probe are non-blocking as well, the result can be assembled without using an executor thread, and thereby avoids the queueing. 

The liveness probe of Keycloak has as default no entries, so it will always be reactive. The readiness probe contains a check which attempts to prepare a non-blocking result when there are active DB connections, and falls back to a blocking check when there are none. 

Let me know if this resolves your question.

Reply:
Thanks for the explanation! I guess we can discuss any remaining questions later. ;)